### PR TITLE
Phase2 L1 NN Tau

### PR DESCRIPTION
**Pull request by rekovic**
#### PR description:

This is effectively the PR #30482 (L1Trigger/Phase2L1ParticleFlow package which has been reviewed fully) augmented by the very last one commit dd150add31b5b7a3a7ea2639aec65cdab36816be which adds the producer for L1 NN Taus, contained in the following five files:

 * L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh
 * L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h
 * L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc
 * L1Trigger/Phase2L1ParticleFlow/python/L1NNTauProducer_cff.py
 * L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc

The NNTau producer is  added to the Phase2 L1T sequence in 
 * L1Trigger/Configuration/python/SimL1Emulator_cff.py

@kpedro88 
Please review those.
 

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by rekovic** [comment]
assign upgrade

**Comment by rekovic** [comment]
assign l1

File: L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+public:
+  explicit L1NNTauProducer(const edm::ParameterSet &);
+  ~L1NNTauProducer();
+
+private:
+  TauNNId *fTauNNId;
+  void addTau(l1t::PFCandidate &iCand,
+              const l1t::PFCandidateCollection &iParts,
+              std::unique_ptr<PFTauCollection> &outputTaus);
+  float deltaR(auto iPart1, auto iPart2);
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh:30]: use existing https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaR.h

File: L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#ifndef L1TRIGGER_PHASE2L1PARTICLEFLOW_L1NNTAU_H
+#define L1TRIGGER_PHASE2L1PARTICLEFLOW_L1NNTAU_H
+
+#include <iostream>
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh:4]: remove unneeded include

File: L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#include <vector>
+#include <TLorentzVector.h>
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh:6]: remove unneeded include

File: L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#ifndef L1TRIGGER_PHASE2L1PARTICLEFLOWS_TAUNNID_H
+#define L1TRIGGER_PHASE2L1PARTICLEFLOWS_TAuNNID_H
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h:2]: inconsistent flags, please check spelling/capitalization

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+L1NNTauProducer::L1NNTauProducer(const edm::ParameterSet& cfg)
+    : fSeedPt_(cfg.getParameter<double>("seedpt")),
+      fConeSize_(cfg.getParameter<double>("conesize")),
+      fTauSize_(cfg.getParameter<double>("tausize")),
+      fMaxTaus_(cfg.getParameter<int>("maxtaus")),
+      fNParticles_(cfg.getParameter<int>("nparticles")),
+      fL1PFToken_(consumes<vector<l1t::PFCandidate> >(cfg.getParameter<edm::InputTag>("L1PFObjects"))) {
+  std::string lNNFile = cfg.getParameter<std::string>("NNFileName");  //,"L1Trigger/Phase2L1Taus/data/tau_3layer.pb");
+  fTauNNId = new TauNNId();
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:11]: memory leak, please use smart pointer
also, member variable should end in underscore

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  if (lNNFile.find("v0") == std::string::npos)
+    fTauNNId->initialize("input_1:0", lNNFile, fNParticles_);
+  if (lNNFile.find("v0") != std::string::npos)
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:14]: else if

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+void L1NNTauProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<l1t::PFCandidateCollection> l1PFCandidates;
+  iEvent.getByToken(fL1PFToken_, l1PFCandidates);
+
+  l1t::PFCandidateCollection pfChargedHadrons;
+  l1t::PFCandidateCollection pfChargedHadrons_sort;
+  l1t::PFCandidateCollection pfChargedHadrons_seeds;
+  for (auto l1PFCand : *l1PFCandidates)
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:26]: const auto&

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    if ((l1PFCand.id() == l1t::PFCandidate::ChargedHadron || l1PFCand.id() == l1t::PFCandidate::Electron) &&
+        fabs(l1PFCand.eta()) < 2.5)
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:28]: std::abs()

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  if (lTaus->size() == 0) {
+    PFTau dummy;
+    lTaus->push_back(dummy);
+  }
+  std::sort(lTaus->begin(), lTaus->end(), [](l1t::PFTau i, l1t::PFTau j) { return (i.pt() > j.pt()); });
+  iEvent.put(std::move(lTaus), "L1PFTausNN");
+}
+
+// create taus based on grid structure
+void L1NNTauProducer::addTau(l1t::PFCandidate& iCand,
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:67]: const auto&

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+      lTaus->push_back(dummy);
+    }
+    iEvent.put(std::move(lTaus), "L1PFTausNN");
+    return;
+  }
+  pfChargedHadrons_seeds.push_back(pfChargedHadrons_sort[0]);
+  for (unsigned int i0 = 1; i0 < pfChargedHadrons_sort.size(); i0++) {
+    bool pMatch = false;
+    for (unsigned int i1 = 0; i1 < pfChargedHadrons_seeds.size(); i1++) {
+      if (deltaR(pfChargedHadrons_seeds[i1], pfChargedHadrons_sort[i0]) < 0.4)
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:46]: should 0.4 be replaced with fConeSize_?

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 11]
```diff
@@ -0,0 +1,77 @@
+#include "L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h"
+#include <iostream>
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:2]: unneeded include

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+#include <cmath>
+#include <TMath.h>
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:4]: unneeded include

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+TauNNId::TauNNId() { NNvectorVar_.clear(); }
+TauNNId::~TauNNId() {
+  tensorflow::closeSession(session_);
+  delete graphDef_;
+}
+void TauNNId::initialize(const std::string &iInput, const std::string &iWeightFile, int iNParticles) {
+  std::string cmssw_base_src = getenv("CMSSW_BASE");
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:12]: better to use FileInPath to get the full path automatically

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+  graphDef_ = tensorflow::loadGraphDef((cmssw_base_src + "/src/" + iWeightFile).c_str());
+  session_ = tensorflow::createSession(graphDef_);
+  fNParticles_ = iNParticles;
+  fPt_ = std::make_unique<float>(fNParticles_);
+  fEta_ = std::make_unique<float>(fNParticles_);
+  fPhi_ = std::make_unique<float>(fNParticles_);
+  fId_ = std::make_unique<float>(fNParticles_);
+  fInput_ = iInput;  //tensorflow::run(session_, { { "input_1:0",input } }, { "dense_4/Sigmoid:0" }, &outputs);
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:20]: delete commented-out code

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+  for (int i0 = 0; i0 < fNParticles_; i0++) {
+    NNvectorVar_.push_back(fPt_.get()[i0]);   //pT
+    NNvectorVar_.push_back(fEta_.get()[i0]);  //dEta from jet axis
+    NNvectorVar_.push_back(fPhi_.get()[i0]);  //dPhi from jet axis
+    if (fPt_.get()[i0] == 0) {
+      for (int i1 = 0; i1 < 5; i1++)
+        NNvectorVar_.push_back(0);
+      continue;
+    }
+    fId_.get()[i0] == l1t::PFCandidate::Photon ? NNvectorVar_.push_back(1) : NNvectorVar_.push_back(0);    //Photon
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:33]: simplify: `NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::Photon);`
(also for lines below)

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    fId_.get()[i0] = 0;
+  }
+  std::sort(iParts.begin(), iParts.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) { return (i.pt() > j.pt()); });
+  for (unsigned int i0 = 0; i0 < iParts.size(); i0++) {
+    if (i0 > 10)
+      break;
+    fPt_.get()[i0] = iParts[i0].pt();
+    fEta_.get()[i0] = iSeed.eta() - iParts[i0].eta();
+    float lDPhi = iSeed.phi() - iParts[i0].phi();
+    if (lDPhi > TMath::Pi())
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:68]: use M_PI or https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaPhi.h

File: L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh (hunk 1 above)
Comment (kpedro88) [reply L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh:30]: (also, abbreviated function templates using auto are a C++20 feature, not supported in CMSSW yet, hence the compilation error in the PR test)

File: L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h (hunk 4 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h:2]: done

File: L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh (hunk 2 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh:4]: done

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc (hunk 6 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:14]: done

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc (hunk 10 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:46]: yes. done

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc (hunk 11 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:2]: done

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc (hunk 12 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:4]: done

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc (hunk 14 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:20]: done

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc (hunk 15 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:33]: done

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc (hunk 16 above)
Comment (rekovic) [reply L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:68]: done

**Comment by rekovic** [comment]
addressed cmments and now rebasing off of #30510

**Comment by rekovic** [comment]
please test

**Comment by silviodonato** [comment]
Since this PR are 10k of new code, I would say to merge them after 11_1_0.

**Comment by rekovic** [comment]
@silviodonato 
It is only 5 new files difference w.r.t. to #30510, which should be very little new code.

**Comment by kpedro88** [comment]
some comments not addressed:
https://github.com/cms-sw/cmssw/pull/30505/files#r449090157
https://github.com/cms-sw/cmssw/pull/30505/files#r449094000

also this PR will probably need to be rebased

**Comment by kpedro88** [comment]
@rekovic what is the status of this PR

**Comment by rekovic** [comment]
@kpedro88 
Addressing the two comments, turning the Tau on, and rebasing to reflect only the 5 new files w.r.t. `cms-sw/master`.
The failing tests of #30521 are passing locally with the external files checked out locally.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

File: L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  std::string fInput_;
+  int fNParticles_;
+  unique_ptr<float> fPt_;
+  unique_ptr<float> fEta_;
+  unique_ptr<float> fPhi_;
+  unique_ptr<float> fId_;
+
+private:
+  tensorflow::Session *session_;
+  tensorflow::GraphDef *graphDef_;
```
Comment (makortel) [inline L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h:27]: How big is the graph? I think we've typically shared the graph across the stream modules in an `edm::GlobalCache`.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+  if (pfChargedHadrons_sort.empty()) {
+    if (lTaus->empty()) {
+      PFTau dummy;
+      lTaus->push_back(dummy);
+    }
+    iEvent.put(std::move(lTaus), "L1PFTausNN");
+    return;
+  }
+  pfChargedHadrons_seeds.push_back(pfChargedHadrons_sort[0]);
+  for (unsigned int i0 = 1; i0 < pfChargedHadrons_sort.size(); i0++) {
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:45]: could be range-based loop

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    bool pMatch = false;
+    for (unsigned int i1 = 0; i1 < pfChargedHadrons_seeds.size(); i1++) {
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:47]: could be range-based loop

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+      if (deltaR(pfChargedHadrons_seeds[i1], pfChargedHadrons_sort[i0]) < fConeSize_)
+        pMatch = true;
+    }
+    if (pMatch)
+      continue;
+    pfChargedHadrons_seeds.push_back(pfChargedHadrons_sort[i0]);
+    if (int(pfChargedHadrons_seeds.size()) > fMaxTaus_ - 1)
+      break;
+  }
+  for (unsigned int i0 = 0; i0 < pfChargedHadrons_seeds.size(); i0++) {
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:57]: could be range-based loop

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+void L1NNTauProducer::addTau(const l1t::PFCandidate& iCand,
+                             const l1t::PFCandidateCollection& iParts,
+                             std::unique_ptr<l1t::PFTauCollection>& outputTaus) {
+  l1t::PFCandidateCollection pfTauCands;
+  TLorentzVector lTot;
+  lTot.SetPtEtaPhiM(0, 0, 0, 0);
+  TLorentzVector lCand;
+  lCand.SetPtEtaPhiM(0, 0, 0, 0);
+  int lId = 0;
+  for (auto l1PFCand : iParts) {
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:78]: const auto&

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+  if (lTot.Pt() < fSeedPt_)
+    return;
+  std::sort(
+      pfTauCands.begin(), pfTauCands.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) { return (i.pt() > j.pt()); });
+  float NN = fTauNNId_->compute(iCand, pfTauCands);
+  math::PtEtaPhiMLorentzVector tempP4(lCand.Pt(), lCand.Eta(), lCand.Phi(), lCand.M());
+  l1t::PFTau l1PFTau(tempP4, NN, 0, lId);
+  outputTaus->push_back(l1PFTau);
+}
+float L1NNTauProducer::deltaR(const l1t::PFCandidate& iPart1, const l1t::PFCandidate& iPart2) {
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:101]: use https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaR.h

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+#include <cmath>
+
+L1NNTauProducer::L1NNTauProducer(const edm::ParameterSet& cfg)
+    : fSeedPt_(cfg.getParameter<double>("seedpt")),
+      fConeSize_(cfg.getParameter<double>("conesize")),
+      fTauSize_(cfg.getParameter<double>("tausize")),
+      fMaxTaus_(cfg.getParameter<int>("maxtaus")),
+      fNParticles_(cfg.getParameter<int>("nparticles")),
+      fL1PFToken_(consumes<vector<l1t::PFCandidate> >(cfg.getParameter<edm::InputTag>("L1PFObjects"))) {
+  std::string lNNFile = cfg.getParameter<std::string>("NNFileName");  //,"L1Trigger/Phase2L1Taus/data/tau_3layer.pb");
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:12]: delete commented-out code

File: L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h (hunk 17 above)
Comment (kpedro88) [reply L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h:27]: also, a smart pointer could be used (though it seems like that isn't common practice for CMSSW TF code at the moment, something to be improved globally)

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    fId_.get()[i0] = 0;
+  }
+  std::sort(iParts.begin(), iParts.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) { return (i.pt() > j.pt()); });
+  for (unsigned int i0 = 0; i0 < iParts.size(); i0++) {
+    if (i0 > 10)
+      break;
+    fPt_.get()[i0] = iParts[i0].pt();
+    fEta_.get()[i0] = iSeed.eta() - iParts[i0].eta();
+    float lDPhi = iSeed.phi() - iParts[i0].phi();
+    if (lDPhi > M_PI)
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:65]: use https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaPhi.h ?

File: L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+float TauNNId::compute(const l1t::PFCandidate &iSeed, l1t::PFCandidateCollection &iParts) {
+  for (int i0 = 0; i0 < fNParticles_; i0++) {
+    fPt_.get()[i0] = 0;
+    fEta_.get()[i0] = 0;
+    fPhi_.get()[i0] = 0;
+    fId_.get()[i0] = 0;
+  }
+  std::sort(iParts.begin(), iParts.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) { return (i.pt() > j.pt()); });
+  for (unsigned int i0 = 0; i0 < iParts.size(); i0++) {
+    if (i0 > 10)
```
Comment (kpedro88) [inline L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc:60]: why is this hardcoded? is it the same as the "nparticles" python parameter? if so, just use that...
(also, just make the loop condition `i0 < std::min(iParts.size(),fNparticles_)`)

**Comment by silviodonato** [comment]
urgent

**Comment by silviodonato** [comment]
@rekovic there is a conflict

**Comment by kpedro88** [comment]
@rekovic there are also still comments to be addressed

**Comment by kpedro88** [comment]
@rekovic what is the status of this PR

**Comment by kpedro88** [comment]
@rekovic ping
(clearly this PR is not actually "urgent"...)

**Comment by kpedro88** [comment]
-1
to remove from upgrade list

**Comment by silviodonato** [comment]
@rekovic please let me know when you want to open this PR back

**Pull request closed without merging**